### ISSUE TITLE
systemctl --user stopped working

### ISSUE THREAD

**Issue by DamienCassou** [NixOS/nixpkgs#11480]
I've just updated to Gnome 3.18 and this is what I get in gnome-terminal:

```
$ systemctl --user
Failed to list units: Process org.freedesktop.systemd1 exited with status 1

$ emacs-service status
Failed to get properties: Process org.freedesktop.systemd1 exited with status 1
```

I've tried to go back to 3.16 but the problem persists. I guess I did something wrong, but I've no idea where.

**Comment by DamienCassou** [comment]
When I say I can't go back to 3.16, I mean I can't build a new configuration after changing back my configuration.nix to use 3.16. But, I can go back to a previous version using the boot menu.

**Comment by domenkozar** [comment]
This is probably due to newer systemd 227.

**Comment by DamienCassou** [comment]
@domenkozar what can I do about this?

**Comment by DamienCassou** [comment]
Someone on IRC #systemd told me that my error message indicates the absence of systemd in the user session. He also told me that pam_systemd might be the cause.

```
$ grep pam_systemd /etc/pam.d/*
/etc/pam.d/gdm:session  optional       /nix/store/jnzaa18j5sqnqdxd0m13fz2sj5rl01f2-systemd-227/lib/security/pam_systemd.so
/etc/pam.d/gdm-autologin:session  optional       /nix/store/jnzaa18j5sqnqdxd0m13fz2sj5rl01f2-systemd-227/lib/security/pam_systemd.so
/etc/pam.d/gdm-launch-environment:session  optional       /nix/store/jnzaa18j5sqnqdxd0m13fz2sj5rl01f2-systemd-227/lib/security/pam_systemd.so
/etc/pam.d/gdm-password:session  optional       /nix/store/jnzaa18j5sqnqdxd0m13fz2sj5rl01f2-systemd-227/lib/security/pam_systemd.so
/etc/pam.d/login:session optional /nix/store/jnzaa18j5sqnqdxd0m13fz2sj5rl01f2-systemd-227/lib/security/pam_systemd.so
/etc/pam.d/sshd:session optional /nix/store/jnzaa18j5sqnqdxd0m13fz2sj5rl01f2-systemd-227/lib/security/pam_systemd.so
/etc/pam.d/systemd-user:session optional /nix/store/jnzaa18j5sqnqdxd0m13fz2sj5rl01f2-systemd-227/lib/security/pam_systemd.so
```

**Comment by DamienCassou** [comment]
Putting in CC the recent authors of `modules/security/pam.nix` which might be the cause of this problem:

/cc @obadz @oxij @offlinehacker @joachifm @philandstuff

**Comment by DamienCassou** [comment]
```
$ journalctl --boot | grep -i pam
déc. 06 09:15:52 luz3 systemd[1]: systemd 227 running in system mode. (+PAM -AUDIT -SELINUX +IMA -APPARMOR -SMACK -SYSVINIT +UTMP -LIBCRYPTSETUP +GCRYPT -GNUTLS +ACL +XZ -LZ4 +SECCOMP +BLKID -ELFUTILS +KMOD -IDN)
déc. 06 09:15:56 luz3 systemd[847]: pam_unix(systemd-user:session): session opened for user gdm by (uid=0)
déc. 06 09:16:22 luz3 gdm-password][1071]: gkr-pam: no password is available for user
déc. 06 09:16:26 luz3 gdm-password][1071]: pam_unix(gdm-password:session): session opened for user cassou by (unknown)(uid=0)
déc. 06 09:16:26 luz3 systemd[1077]: pam_unix(systemd-user:session): session opened for user cassou by (uid=0)
déc. 06 09:16:26 luz3 systemd[1078]: pam_unix(systemd-user:session): session closed for user cassou
déc. 06 09:16:47 luz3 systemd[848]: pam_unix(systemd-user:session): session closed for user gdm
déc. 06 09:30:32 luz3 pkexec[2046]: pam_unix(polkit-1:session): session opened for user root by (uid=1000)
déc. 06 09:30:33 luz3 pkexec[2053]: pam_unix(polkit-1:session): session opened for user root by (uid=1000)
déc. 06 09:30:33 luz3 pkexec[2059]: pam_unix(polkit-1:session): session opened for user root by (uid=1000)
déc. 06 09:30:33 luz3 pkexec[2065]: pam_unix(polkit-1:session): session opened for user root by (uid=1000)
déc. 06 09:30:34 luz3 pkexec[2071]: pam_unix(polkit-1:session): session opened for user root by (uid=1000)
déc. 06 09:30:35 luz3 pkexec[2077]: pam_unix(polkit-1:session): session opened for user root by (uid=1000)
déc. 06 09:53:04 luz3 gdm-password][2517]: gkr-pam: no password is available for user
déc. 06 09:53:04 luz3 gdm-password][2517]: pam_unix(gdm-password:auth): authentication failure; logname= uid=0 euid=0 tty= ruser= rhost=  user=cassou
déc. 06 09:53:08 luz3 gdm-password][2521]: gkr-pam: no password is available for user
```

**Comment by hrdinka** [comment]
This might be related to https://github.com/systemd/systemd/issues/1507 as well. I am currently suffering from this bug. What it does is that it tries to unmount all mounted filesystems as soon as a user session is started. If you log in as root it will actually umount stuff, also systemds filesystems under `/run` (setuid wrappers for example). This leaves my computer in a semi broken state where many operations won't work. When trying to substitute user via `su` I get similiar error messages to yours.

Have a look at your journal after a fresh start and look for umounting errors. Currently I am running systemd version 228 without this problem.

**Comment by oxij** [comment]
Christoph Hrdinka <notifications@github.com> writes:

> This might be related to
> https://github.com/systemd/systemd/issues/1507 as well. I am currently
> suffering from this bug. What it does is that it tries to unmount all
> mounted filesystems as soon as a user session is started. If you log
> in as root it will actually umount stuff, also systemds filesystems
> under `/run` (setuid wrappers for example). This leaves my computer in
> a semi broken state where many operations won't work. When trying to
> substitute user via `su` I get similiar error messages to yours.

Do you by any chance have

    programs.ssh.startAgent = false;

in your `configuration.nix`?

**Comment by dezgeg** [comment]
This is sort-of getting offtopic wrt. the original `--user` issue, but I've noticed the everything-getting-unmounted problem on 227 as well, and 228 indeed fixed it.

**Comment by DamienCassou** [comment]
@oxij indeed, I use gnupg as ssh-agent so `programs.ssh.startAgent = false` for me. What does it change?

**Comment by oxij** [comment]
Me too, and I noticed that I had this unmount-everything problem only on
machines with that line in `configuration.nix`. I was lazy to debug it,
so I simply removed it and it worked. It might have coincided with
systemd update, though, but it's a food for thought.

**Comment by hrdinka** [comment]
The only machine where I have this problem is the one with `programs.ssh.startAgent = false`. However this might be a workaround but systemd 227 still is broken.

@DamienCassou do you have unmount errors in your journal? Once systemd unmounted everything I wasn't able to `su` nor to `systemctl --user` (cause it unmounts /run/user/* as well). That is why I don't think that this is offtopic @dezgeg

**Comment by dezgeg** [comment]
Oh, I wasn't aware that systemd stores stuff in /run/user but indeed that's where some control sockets seem to be lying around!

So a quick way to try the new systemd would be `curl https://github.com/dezgeg/nixpkgs/commit/1f9cea944c5811cb2be374b9bb4c2403463998fc.patch | git am` on top of whatever nixpkgs commit you're currently running. (https://github.com/dezgeg/systemd/tree/nixos-v228 is just a trivial rebase of nixos-v227 to the new version)

**Comment by DamienCassou** [comment]
@hrdinka 

```
$ journalctl --boot | grep unmount
déc. 06 14:59:04 luz3 systemd[5833]: Failed unmounting /home.
déc. 06 14:59:04 luz3 systemd[5833]: Failed unmounting /dev/hugepages.
déc. 06 14:59:04 luz3 systemd[5833]: Failed unmounting /run/keys.
[...]
```

**Comment by hrdinka** [comment]
@DamienCassou yep this looks like the problem. Have you already tried rebuilding your system with either `programs.ssh.startAgent = true` or systemd 228? If one of these fixes gnome for you we might have found the problem for that. Then we should rename the issue to something like systemd 227 is broken and start working on that one.

I am however glad that I am not the only one with this problem. I first thought that something on my computer was broken cause I didn't had this issue on any of my servers. Currently I am trapped in an unfulfilling chain where I do recompile my whole system for a day just to find out that something else fails afterwards and I do have to start over again.

Maybe others are having problems too and haven't noticed it yet. Although this bugs implications are pretty severe most of its effect does only show in small details all around the place (as long as nothing as big as your desktop fails). For me starting X and getty is delayed, setuids arent working (`su`, `ping`) and nixos-containers seem to broken as well (not sure if this is connected to this).

**Comment by DamienCassou** [comment]
@hrdinka I applied the patch and my computer is building. But it looks like every single package is getting compiled. Does everything depend on systemd?

**Comment by hrdinka** [comment]
That's the nasty endless cycle of compilation I was talking about ☺

**Comment by peti** [comment]
Is anyone aware of a compelling reason not to merge https://github.com/dezgeg/nixpkgs/commit/1f9cea944c5811cb2be374b9bb4c2403463998fc into `master`? This bug is really bad, and it needs to be fixed soon. That patch seems to fix the bug, so I'd be in favor or making it available to our users.

**Comment by hrdinka** [comment]
@peti :+1:

**Comment by domenkozar** [comment]
@peti 

1) it's a mass-rebuild as far as I'm aware
2) it needs to be merged into NixOS/systemd repo

**Comment by peti** [comment]
The change causes significant re-builds, but it's not a stdenv change AFAIC. Without this patch, a lot of people's installations are badly broken, and IMHO the benefit of fixing that outweighs the costs of the re-builds.

Merging the changes into the `NixOS/systemd` repository would be nice, of course, but there's no technical reason why it needs to be done before we can do the update. (Unless I'm missing anything?) We can update the NixOS variant of that repo later and switch back to it -- and we should! -- but ultimately the question of whether the codes lives in place *A* or *B* feels like a technicality that doesn't matter so much.

**Comment by domenkozar** [comment]
@peti agreed, but I'd like @edolstra to take a look and meanwhile update our "official" repo while he's at it. If he doesn't have time, I agree we should @peti  plan.

**Comment by hrdinka** [comment]
This bug is pretty severe and many people who might not have noticed it yet may be affected by it. And we really should fix it before they notice it the hard way. Maybe many issues users have refer to this one. @peti for example stumbled upon it by accident because of a grub fail https://github.com/NixOS/nixpkgs/issues/11521. I noticed it because `su` stopped working. @DamienCassou (still waiting for confirmation) isn't able to login. This bug is hard to track down any many users might be looking for a solution for their weird problem right now.

I am kinda horrified by the idea of having this bug triggered on a server machine. It will unmount `/run/keys` (not nice if you deploy all your secrets via `nixops`) and unmount `/run/user/*` likely rendering your system to a brick that cannot be accessed via ssh any more. Not mentioning terminal/VNC or any service using setuid programs.

I don't mind the many rebuilds caused by it, because if you are building from master you are already rebuilding your whole system every two weeks due the vast amount of changes.

**Comment by peti** [comment]
@domenkozar definitely has a point that it's good style to involve @edolstra in this update. He's the one who put most effort into systemd so far and he's most qualified to judge whether this update may have side-effects that end up being worse that the issue we're trying to fix. Let's give it a day or two so that Eelco has a chance to weigh in, and if it turns out that he can't review this patch for some reason, then we'll merge it without him and hope for the best.

**Comment by edolstra** [comment]
Fixed by 1da87d40628c99081bac0a5d8b32c27123889826.

**Comment by hrdinka** [comment]
@peti I am sorry, that wasn't the message I wanted to deliver. I didn't want to rush nor circumvent @edolstra. I would have asked him for review and help as soon as @DamienCassou got back with results anyway. I just wanted to point out that it is more severe than one without the problem might think and that we should keep this bug in mind when we are helping others with their issues. Just as with your grub issue making a connection to this one isn't easy and might cost many evenings of debugging.

**Comment by DamienCassou** [comment]
After all this time compiling every single byte on my system, I'm happy to report that this patch, updating systemd to 228, fixed my problem. Thank you very much

**Comment by domenkozar** [comment]
channel has updated, do test :)
